webserver/events/broker.go

```
@@ -0,0 +1,246 @@
+	clients *list.List
+	// max number of clients per a broker instance
+	maxClientsCount uint
+
+	// ConnectionChan is a shared queue to buffer incoming websocket connections
+	// closing connChan will terminate the broker
+	ConnectionChan chan wsConn
+
+	// Events
+	acceptedBlockChan chan block.Block
+	listener          eventbus.TopicListener
+	// topics.Tx acceptedTxChan chan transactions.Transaction
+}
+
+func NewBroker(id uint, eventBus *eventbus.EventBus, maxClientsCount uint, connChan chan wsConn) *Broker {
+
+	b := new(Broker)
+	b.ConnectionChan = connChan
+	b.acceptedBlockChan, b.listener = consensus.InitAcceptedBlockUpdate(eventBus)
+	b.clients = list.New()
+	b.maxClientsCount = maxClientsCount
+	b.id = id
+	return b
+}
+
+func (b *Broker) Run() {
+
+Loop:
+	for {
+		// Any of the handlers must be capable of recovering from panic
+		select {
+		// new client connection from webserver
+		case conn, isOpen := <-b.ConnectionChan:
+			if isOpen {
+				b.handleconn(conn)
+			} else {
+				// Terminate the broker when the shared connChan is closed
+				break Loop
+			}
+		// new accepted block from node
+		case blk := <-b.acceptedBlockChan:
+			b.handleblock(blk)
+		case <-time.After(30 * time.Second):
+			b.handleIdle()
+		}
+	}
+
+	// Teardown procedure.
+
+	// It's advisable to reset the broker state entirely.
+	// This would allow later to restart a broker without any leaks.
+	log.Infof("Terminating broker %d", b.id)
+
+	// Unsubscribe from all eventBus events
+	b.listener.Quit()
+
+	// Terminate all clients goroutines
+	for e := b.clients.Front(); e != nil; e = e.Next() {
+		c := e.Value.(*wsClient)
+		close(c.msgChan)
+	}
+
+	// reset clients list
+	b.clients.Init()
+}
+
+// handleblock handles the topics.AcceptedBlock event emitted from node layer It
+// packs a json from the block and broadcast it to all active clients
+func (b *Broker) handleblock(blk block.Block) {
+
+	defer func() {
+		if r := recover(); r != nil {
```

Not sure I understand why you have `recover`s in place for this and `handleconn`, but I suspect it has to do with the closing of channels? If so, due to the channel-centric design of the broker, I think these might be superfluous.

---

No, recover is here to keep the goroutine running even after a panic error.  Broker routine should not get down silently for any reason.

---

Just to clarify, if all broker routines gets down, explore would become useless and still running without any indication. Later on, we can add restart-on-panic  capability for the broker routines. Once we have the `watchdog` routine implemented,  `recover()` would be superfluous.

---

> No, recover is here to keep the goroutine running even after a panic error.

I get that, but I'm trying to see if that is really needed. At first glance I don't see anything that's a hazard for panics.

---

> Just to clarify, if all broker routines get down, block explore would become useless and still running without any indication for failure.

Just to be clear, doesn't a `panic` crash an entire process?

---

"Recover is a built-in function that regains control of a panicking goroutine. Recover is only useful inside deferred functions. During normal execution, a call to recover will return nil and have no other effect. If the current goroutine is panicking, a call to recover will capture the value given to panic and resume normal execution. "

source: https://blog.golang.org/defer-panic-and-recover

---

I understand what `recover` does; again, I am trying to see exactly why it is here because I don't see anything that could result in a panic very easily.

---

The above answers your question. `doesn't a panic crash an entire process?` - it will not crash the entire process if recover() is called from any of the functions up in the stack of the current goroutine.

WRT `I don't see anything that could result in a panic very easily.`

 It's for runtime errors that could appear with current impl or future changes.

---

Well alright. Personally I think it would make more sense to let the process crash and then applying a patch for whatever problem arises, but if you really deem it necessary to keep things running in this case, then it's fine.

> Broker routine should not get down silently for any reason.
> Just to clarify, if all broker routines get down, block explore would become useless and still running without any indication for failure.

This is exactly why I said `doesn't a panic crash an entire process?`. Earlier, you seemed to be alluding to `recover` being a fix for the silent shutdown of goroutines, though I think `recover` could actually give way to something similar to it, as if there was any runtime error that prevents that goroutine from taking any more action, it will simply keep recovering and no longer be useful. Maybe, I am misunderstanding your points.

---

Maybe my explanation is not best. Sorry for that. You're right the process would panic and not goroutine silently terminating.

The point with the recover() is not about bugs but about runtime errors for each it's cheaper to recover from instead of terminating the process. 

One example situation:
E.g eventBus sends a few blocks with nil Header for any reason (bug in eventbus), this will make  the handler panic and the whole process  gets down instead just recover from it and continue with next events.

---

Another example is that we don't utilize recover() in peer go-routines in dusk-blockchain. That allows invalid message to terminate the entire process instead of just recover from the malicious P2P message.